File: benchmarks/R/benchmark_twfe.R
#!/usr/bin/env Rscript
# Benchmark: Two-Way Fixed Effects (R `fixest` package with absorbed FE)
#
# This uses fixest::feols() with absorbed unit + post FE and unit-level clustering,
# matching the Python TwoWayFixedEffects estimator's approach.
#
# Usage:
#   Rscript benchmark_twfe.R --data path/to/data.csv --output path/to/results.json

library(fixest)
library(jsonlite)
library(data.table)

# Parse command line arguments
args <- commandArgs(trailingOnly = TRUE)

parse_args <- function(args) {
  result <- list(
    data = NULL,
    output = NULL
  )

  i <- 1
  while (i <= length(args)) {
    if (args[i] == "--data") {
      result$data <- args[i + 1]
      i <- i + 2
    } else if (args[i] == "--output") {
      result$output <- args[i + 1]
      i <- i + 2
    } else {
      i <- i + 1
    }
  }

  if (is.null(result$data) || is.null(result$output)) {
    stop("Usage: Rscript benchmark_twfe.R --data <path> --output <path>")
  }

  return(result)
}

config <- parse_args(args)

# Load data
message(sprintf("Loading data from: %s", config$data))
data <- fread(config$data)

# Run benchmark
message("Running TWFE estimation with absorbed FE...")
start_time <- Sys.time()

# TWFE with absorbed unit + post fixed effects, clustered at unit level
# This matches Python's TwoWayFixedEffects:
#   - Within-transformation removes unit and time (post) FE
#   - Cluster-robust SE at unit level (automatic)
model <- feols(
  outcome ~ treated:post | unit + post,
  data = data,
  cluster = ~unit
)

estimation_time <- as.numeric(difftime(Sys.time(), start_time, units = "secs"))

# Extract results
coef_name <- "treated:post"
coefs <- coef(model)
ses <- se(model)
pvals <- pvalue(model)
ci <- confint(model)

# Find the treatment effect coefficient
if (coef_name %in% names(coefs)) {
  att <- coefs[coef_name]
  att_se <- ses[coef_name]
  att_pval <- pvals[coef_name]
  att_ci <- ci[coef_name, ]
} else {
  # Try alternative name formats
  idx <- grep("treated.*post|post.*treated", names(coefs))
  if (length(idx) > 0) {
    att <- coefs[idx[1]]
    att_se <- ses[idx[1]]
    att_pval <- pvals[idx[1]]
    att_ci <- ci[idx[1], ]
    coef_name <- names(coefs)[idx[1]]
  } else {
    stop("Could not find treatment effect coefficient")
  }
}

# Format output
results <- list(
  estimator = "fixest::feols (absorbed FE)",
  cluster = "unit",

  # Treatment effect
  att = unname(att),
  se = unname(att_se),
  pvalue = unname(att_pval),
  ci_lower = unname(att_ci[1]),
  ci_upper = unname(att_ci[2]),
  coef_name = coef_name,

  # Model statistics
  model_stats = list(
    r_squared = summary(model)$r2,
    adj_r_squared = summary(model)$adj.r2,
    n_obs = model$nobs
  ),

  # Timing
  timing = list(
    estimation_seconds = estimation_time,
    total_seconds = estimation_time
  ),

  # Metadata
  metadata = list(
    r_version = R.version.string,
    fixest_version = as.character(packageVersion("fixest")),
    n_units = length(unique(data$unit)),
    n_periods = length(unique(data$post)),
    n_obs = nrow(data)
  )
)

# Write output
message(sprintf("Writing results to: %s", config$output))
write_json(results, config$output, auto_unbox = TRUE, pretty = TRUE, digits = 15)

message(sprintf("Completed in %.3f seconds", estimation_time))

File: benchmarks/python/benchmark_twfe.py
#!/usr/bin/env python3
"""
Benchmark: TwoWayFixedEffects (diff-diff TwoWayFixedEffects class).

This benchmarks the actual TwoWayFixedEffects class with within-transformation,
as opposed to benchmark_basic.py which uses DifferenceInDifferences with formula.

Usage:
    python benchmark_twfe.py --data path/to/data.csv --output path/to/results.json
"""

import argparse
import json
import os
import sys
from pathlib import Path

# IMPORTANT: Parse --backend and set environment variable BEFORE importing diff_diff
def _get_backend_from_args():
    """Parse --backend argument without importing diff_diff."""
    parser = argparse.ArgumentParser(add_help=False)
    parser.add_argument("--backend", default="auto", choices=["auto", "python", "rust"])
    args, _ = parser.parse_known_args()
    return args.backend

_requested_backend = _get_backend_from_args()
if _requested_backend in ("python", "rust"):
    os.environ["DIFF_DIFF_BACKEND"] = _requested_backend

# NOW import diff_diff and other dependencies (will see the env var)
import pandas as pd

# Add parent to path for imports
sys.path.insert(0, str(Path(__file__).parent.parent.parent))

from diff_diff import TwoWayFixedEffects, HAS_RUST_BACKEND
from benchmarks.python.utils import Timer


def parse_args():
    parser = argparse.ArgumentParser(description="Benchmark TwoWayFixedEffects estimator")
    parser.add_argument("--data", required=True, help="Path to input CSV data")
    parser.add_argument("--output", required=True, help="Path to output JSON results")
    parser.add_argument(
        "--backend", default="auto", choices=["auto", "python", "rust"],
        help="Backend to use: auto (default), python (pure Python), rust (Rust backend)"
    )
    return parser.parse_args()


def get_actual_backend() -> str:
    """Return the actual backend being used based on HAS_RUST_BACKEND."""
    return "rust" if HAS_RUST_BACKEND else "python"


def main():
    args = parse_args()

    actual_backend = get_actual_backend()
    print(f"Using backend: {actual_backend}")

    # Load data
    print(f"Loading data from: {args.data}")
    data = pd.read_csv(args.data)

    # Run benchmark using TwoWayFixedEffects (within-transformation approach)
    print("Running TWFE estimation...")

    twfe = TwoWayFixedEffects(robust=True)  # auto-clusters at unit level

    with Timer() as timer:
        results = twfe.fit(
            data,
            outcome="outcome",
            treatment="treated",
            time="post",
            unit="unit",
        )

    att = results.att
    se = results.se
    pvalue = results.p_value
    ci = results.conf_int

    total_time = timer.elapsed

    # Build output
    output = {
        "estimator": "diff_diff.TwoWayFixedEffects",
        "backend": actual_backend,
        "cluster": "unit",
        # Treatment effect
        "att": float(att),
        "se": float(se),
        "pvalue": float(pvalue),
        "ci_lower": float(ci[0]),
        "ci_upper": float(ci[1]),
        # Model statistics
        "model_stats": {
            "n_obs": len(data),
            "n_units": len(data["unit"].unique()),
            "n_periods": len(data["post"].unique()),
        },
        # Timing
        "timing": {
            "estimation_seconds": total_time,
            "total_seconds": total_time,
        },
        # Metadata
        "metadata": {
            "n_units": len(data["unit"].unique()),
            "n_periods": len(data["post"].unique()),
            "n_obs": len(data),
        },
    }

    # Write output
    print(f"Writing results to: {args.output}")
    output_path = Path(args.output)
    output_path.parent.mkdir(parents=True, exist_ok=True)
    with open(output_path, "w") as f:
        json.dump(output, f, indent=2)

    print(f"Completed in {total_time:.3f} seconds")
    return output


if __name__ == "__main__":
    main()
